tensorflow_probability/python/math/special.py

```diff
@@ -88,280 +88,378 @@ def atan_difference(x, y, name=None):
+    new_d = tf.where(tf.abs(new_d) < small, small, new_d)
+    new_d = tf.math.reciprocal(new_d)
+    delta = new_c * new_d
+    new_f = f * delta
+
+    new_c_grad = (numerator_grad * c - numerator * c_grad) / (c * c)
+    new_d_grad = -new_d * new_d * (numerator_grad * d + numerator * d_grad)
+    new_f_grad = f_grad * delta + (f * new_c_grad * new_d) + (
+        f * new_d_grad * new_c)
+
+    should_stop = should_stop | (tf.math.abs(delta - one) < tolerance)
+    values = (new_c, new_d, new_f)
+    gradients = (new_c_grad, new_d_grad, new_f_grad)
+
+    return should_stop, iteration + one, values, gradients
+
+  # Assume all input Tensors have the same shape. The extra dimension is
+  # needed to compute the gradients with respect to a and b.
+  a, b, x, where = (z[..., tf.newaxis] for z in (a, b, x, where))
+
+  apb = a + b
+  ap1 = a + one
+
+  initial_c = tf.ones_like(x)
+  initial_d = one - apb * x / ap1
+  initial_d = tf.where(tf.abs(initial_d) < small, small, initial_d)
+  initial_d = tf.math.reciprocal(initial_d)
+  initial_f = initial_d
+  initial_values = (initial_c, initial_d, initial_f)
+
+  initial_d_grad = tf.concat([one - b, ap1], axis=-1) * x / tf.square(
+      x * apb - ap1)
+  initial_c_grad = tf.zeros_like(initial_d_grad)
+  initial_f_grad = initial_d_grad
+  initial_gradients = (initial_c_grad, initial_d_grad, initial_f_grad)
+
+  (_, _, values, gradients) = tf.while_loop(
+      cond=lambda stop, *_: tf.reduce_any(~stop),
+      body=continued_fraction_evaluation,
+      loop_vars=(
+          ~where,
+          tf.constant(2., dtype=dtype),
+          initial_values,
+          initial_gradients),
+      maximum_iterations=max_iterations)
+
+  # Remove the previously added extra dimension: it is no longer needed.
+  f = tf.squeeze(values[-1], axis=-1)
+  f_grad_a, f_grad_b = tf.unstack(gradients[-1], axis=-1)
+
+  return f, f_grad_a, f_grad_b
+
+
+def _betainc_der_continued_fraction(a, b, x, dtype, where):
+  """Returns the partial derivatives of betainc with respect to a and b."""
+  # This function evaluates betainc(a, b, x) by its continued fraction
+  # expansion given here: https://dlmf.nist.gov/8.17.E22
+  # We apply this function when the input (a, b, x) does not belong to the
+  # proper region of computation of `_betainc_der_power_series`.
+  one = tf.constant(1., dtype=dtype)
+  two = tf.constant(2., dtype=dtype)
+
+  # This continued fraction expansion of betainc converges rapidly
+  # for x < (a - 1) / (a + b - 2). For x >= (a - 1) / (a + b - 2),
+  # we can obtain an equivalent computation by using the symmetry
+  # relation given here: https://dlmf.nist.gov/8.17#E4
+  apply_symmetry = (x >= (a - one) / (a + b - two))
+  a_orig = a
+  a = tf.where(apply_symmetry, b, a)
+  b = tf.where(apply_symmetry, a_orig, b)
+  x = tf.where(apply_symmetry, one - x, x)
+
+  cf, cf_grad_a, cf_grad_b = _betainc_modified_lentz_method(
+      a, b, x, dtype, where)
+
+  more_terms = tf.math.exp(
```

Comment:
Rename? Maybe normalization? Here and elsewhere.

Reply:
Done.